mm2src/coins/lp_coins.rs

```diff
@@ -2150,15 +2158,44 @@ impl CoinsContext {
         })))
     }
 
-    pub async fn add_coin(&self, coin: MmCoinEnum) -> Result<(), MmError<CoinIsAlreadyActivatedErr>> {
+    pub async fn add_token(&self, coin: MmCoinEnum) -> Result<(), MmError<CoinIsAlreadyActivatedErr>> {
         let mut coins = self.coins.lock().await;
         if coins.contains_key(coin.ticker()) {
             return MmError::err(CoinIsAlreadyActivatedErr {
                 ticker: coin.ticker().into(),
             });
         }
+        let ticker = coin.ticker();
 
-        coins.insert(coin.ticker().into(), coin);
+        let mut platform_coin_tokens = self.platform_coin_tokens.lock();
+        // Here, we tried to add to a token to platform_coin_tokens if the token belongs to a platform coin.
+        if let Some(platform) = platform_coin_tokens.get_mut(coin.platform_ticker()) {
+            platform.push(ticker.to_owned());
+        }
+
+        coins.insert(ticker.into(), coin);
+        Ok(())
+    }
+
+    pub async fn add_l2(&self, coin: MmCoinEnum) -> Result<(), MmError<CoinIsAlreadyActivatedErr>> {
+        let mut coins = self.coins.lock().await;
+        if coins.contains_key(coin.ticker()) {
+            return MmError::err(CoinIsAlreadyActivatedErr {
+                ticker: coin.ticker().into(),
+            });
+        }
+        let ticker = coin.ticker().to_string();
+        let platform_ticker = coin.platform_ticker().to_string();
+        let mut platform_coin_tokens = self.platform_coin_tokens.lock();
+        if let Some(tokens) = platform_coin_tokens.get_mut(&platform_ticker) {
+            if !tokens.contains(&ticker) {
+                tokens.push(ticker.clone());
+            };
+        } else {
+            platform_coin_tokens.insert(platform_ticker, vec![ticker.clone()]);
+        };
```

Comment:
> did you mean that we should return an error if platform_coin_tokens doesn't contain it yet?

I meant that this should be the difference between `add_l2` and `add_token` since `add_token` is used only when activating a platform coin with tokens unlike activating l2 where we activate the platform coin in a separate call first. As a consequence, the platform coin key will not be found in `platform_coin_tokens` when adding an L2 coin.

> In my opinion, it's better to check every place where we put new coins to CoinsContext::coins container, and put the ticker to platform_coin_tokens if coin.ticker() == coin.platform_ticker().

I agree, this solves the above problem.

Reply:
fixed